### DISCUSSION TITLE
【新提案】增加返回值更有意义的set.add、map.set方法

### DISCUSSION THREAD

**Discussion by LongTengDao** [JSCIG/es-discuss#64] [Ideas]
返回`this`的方法过于鸡肋，但修改旧有api不可能了，不过突然想到有没有可能增加新的方法？

## `map.set`

清爽的对象写法：

```js
return obj[key] ??= f();
```

现有的`map.set`写法：

```js
let value = map.get(key);
value ?? map.set(key, value = f());
return value;
```

`map.setValue`提案写法：

```js
return map.get(key) ?? map.setValue(key, f());
```

## `set.add`

清爽的删除写法：

```js
if ( set.delete(value) ) { logic(); }
```

现有的新增写法：

```js
if ( !set.has(value) ) {
    set.add(value);
    initLogic();
}
```

或：

```js
if ( set.size!==set.add(value).size ) {
    initLogic();
}
```

不光啰嗦，性能还差。

`set.addValue`提案写法：

```js
if ( set.addValue(value) ) {
    initLogic();
}
```

方法名可以再讨论。

**Comment by ljharb** [comment]
Is this the same as https://github.com/tc39/proposal-upsert ?

**Reply by LongTengDao** [reply]
Seems better for map.

What about use case 2 (for set)?

**Reply by ljharb** [reply]
I suppose that would require a new method, but you'd need lots of compelling use cases.
